Switched RTH Preset Altitude Override

breadoven opened the pull request:
RTH altitude mode can't be changed in flight which can result in the craft climbing unnecessarily when switched RTH is used to simply bring it home as directly as possible with the pilot able to monitor safe altitude manually. This can be an issue with Max  or Fixed settings causing large climbs for instance when a level route back would have been better.

This PR provides a function that allows the pilot to override the preset RTH altitude during switched RTH by using the AltHold mode switch to effectively reset the RTH altitude to Current during the climb phase, i.e. RTH altitude is set at the altitude achieved when the AltHold switch is selected ON. 

The AltHold mode switch must be set OFF after RTH is initiated before the override function will work to avoid inadvertent override if AltHold switch was already ON. The function is inhibited for failsafe RTH and controlled by a setting, default set to OFF.

The RTH altitude setting returns to the preset value when RTH is switched OFF with the exception of the Max setting which resets to the current altitude.

Been tested on a fixed wing and works as expected.

Related to https://github.com/iNavFlight/inav/issues/6128.

breadoven committed:
Fix Naming

breadoven committed:
Merge branch 'master' into RTH-Altitude-Override

breadoven committed:
Fix Merge Correction Errors

digitalentity commented on src/main/navigation/navigation.c:
```diff
@@ -1152,9 +1157,29 @@ static navigationFSMEvent_t navOnEnteringState_NAV_STATE_RTH_INITIALIZE(navigati
+        return;
+    }
+
+    if (!IS_RC_MODE_ACTIVE(BOXNAVALTHOLD)) {
```
This not-logic is counter-intuitive. Please reverse.

breadoven replied:
Ah OK, yes that's obvious now you mention it. I'll also move the function to after `void updateHomePosition(void) ` (around line 2500 ) so it doesn't break up the sequence of those RTH phases if that's OK with you.

breadoven committed:
Change Control to Pitch Stick

Override trigger changed from AlfHold mode to pitch stick input.

breadoven committed:
Update navigation.c

Fix tabs

breadoven committed:
Update to add RTH Climb First override

RTH Climb First override added using roll stick.
RTH altitude override - pitch down stick
RTH Climb First override - roll right stick
Stick hold for > 2 seconds.

breadoven committed:
Update settings.yaml

breadoven committed:
Update navigation.c

Fix ABS build error

breadoven committed:
Improved Code

Cleaned up code removing Climb First flag. Tested in flight, works as expected.

breadoven committed:
Merge branch 'master' into RTH-Altitude-Override

breadoven committed:
Merge branch 'master' into RTH-Altitude-Override

breadoven committed:
Update Settings.md

breadoven committed:
Fix white space

breadoven committed:
Merge branch 'master' into RTH-Altitude-Override

breadoven committed:
Merge remote-tracking branch 'upstream/master' into RTH-Altitude-Override

breadoven committed:
Update settings.yaml

breadoven committed:
Merge remote-tracking branch 'upstream/master' into RTH-Altitude-Override

breadoven committed:
Update navigation.c

breadoven committed:
Update navigation.c

Reduced stick hold delay time to 1s and limited climb first override to fixed wing only

breadoven committed:
Update Settings
